### PR TITLE
`create.poolpair` `update.poolpair`

### PR DESCRIPTION
**Pull request by monstrobishi**
#### What kind of PR is this?:

/kind feature

#### What this PR does / why we need it:

Index create poolpair and update poolpair.

Note: TODO customRewards, ownerAddress, this is outside the scope of this ticket/PR

#### Which issue(s) does this PR fixes?:

Fixes #

### CODE REVIEW

**Comment by fuxingloh** [comment]
I will review this after TokenIndexer has been merged.

**Comment by monstrobishi** [comment]
Will fix conflicts

File: src/module.model/poolpair.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    id: number // numerical id
+    symbol: string
+  }
+  tokenB: {
+    id: number // numerical id
+    symbol: string
+  }
+  block: {
+    hash: string
+    height: number
```
Comment (fuxingloh) [inline src/module.model/poolpair.ts:80]: Maybe should keep consistency and all the usual block details?

File: src/module.model/poolpair.token.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+export interface PoolPairToken extends Model {
+  id: string // ---------| tokenA-tokenB-poolPairId
+  key: string // --------| tokenA-tokenB
+  poolpairId: number
+
+  block: {
+    hash: string
+    height: number
+  }
```
Comment (fuxingloh) [inline src/module.model/poolpair.token.ts:57]: same for this?

File: src/module.indexer/model/dftx/create.poolpair.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        block: { hash: block.hash, height: block.height, medianTime: block.mediantime, time: block.time }
+      })
+    }
+  }
+
+  async invalidate (block: RawBlock, txns: Array<DfTxTransaction<PoolCreatePair>>): Promise<void> {
+    for (const { dftx: { data } } of txns) {
+      const tokenId = await this.tokenMapper.getNextTokenID(true)
+      await this.poolPairMapper.delete(`${tokenId - 1}-${block.height}`)
+      await this.poolPairTokenMapper.delete(`${data.tokenA}-${data.tokenB}`)
```
Comment (fuxingloh) [inline src/module.indexer/model/dftx/create.poolpair.ts:87]:
```suggestion
      await this.poolPairTokenMapper.delete(`${data.tokenA}-${data.tokenB}-${tokenId}`)
```

**Review by fuxingloh — CHANGES_REQUESTED** [review]
you should add invalidation tests

File: src/module.indexer/model/dftx/create.poolpair.ts (hunk 3 above)
Comment (monstrobishi) [reply src/module.indexer/model/dftx/create.poolpair.ts]: good spotting! actually should be `${data.tokenA}-${data.tokenB}-${tokenId - 1}`, fixed

**Review by fuxingloh — APPROVED** [review]

**Pull request merged**